### PR TITLE
Vectorization worker full pipeline

### PR DESCRIPTION
**Pull request by ciprianjichici**
# Vectorization worker full pipeline

## The issue or feature being addressed

N/A

## Details on the issue fix or feature implementation

Implements the core vectorization pipeline.

## Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [ ]  I have included unit tests for the issue/feature
- [x]  I have included inline docs for my changes, where applicable
- [x]  I have successfully run a local build
- [ ]  I have provided the required update scripts, where applicable
- [x]  I have updated relevant docs, where applicable

> [!NOTE]
> Instead of adding `X`'s inside the checkboxes you wish to check above, first submit the PR, then check the boxes in the rendered description.

### CODE REVIEW

File: src/dotnet/Vectorization/Models/VectorizationState.cs [hunk 1]
```diff
@@ -75,5 +75,12 @@ public class VectorizationState
         public void LogHandlerError(IVectorizationStepHandler handler, string requestId, Exception ex) =>
             LogEntries.Add(new VectorizationLogEntry(
                 requestId, handler.StepId, $"ERROR: {ex.Message}"));
+
+        public static VectorizationState FromRequest(VectorizationRequest request) =>
```
Comment (joelhulen) [inline src/dotnet/Vectorization/Models/VectorizationState.cs:79]: Missing XML documentation on this public method.

File: src/dotnet/Vectorization/Services/RequestManagerService.cs [hunk 2]
```diff
@@ -76,8 +75,8 @@ public async Task Run()
                         // thread pool thread, with no user code higher on the stack (for details, see
```
Comment (joelhulen) [inline src/dotnet/Vectorization/Services/RequestManagerService.cs:75]: Casing

File: src/dotnet/Vectorization/Services/RequestSources/MemoryRequestSourceService.cs [hunk 3]
```diff
@@ -4,49 +4,47 @@
 using FoundationaLLM.Vectorization.Interfaces;
 using System.Collections.Concurrent;
 using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Primitives;
+using FoundationaLLM.Vectorization.Models.Configuration;
 
 namespace FoundationaLLM.Vectorization.Services.RequestSources
 {
-    public class MemoryRequestSourceService : IRequestSourceService
+    public class MemoryRequestSourceService(
```
Comment (joelhulen) [inline src/dotnet/Vectorization/Services/RequestSources/MemoryRequestSourceService.cs:12]: Missing XML documentation. Could probably `<inheritdoc/>`.

File: src/dotnet/Vectorization/Services/RequestSources/RequestSourcesBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+using Microsoft.Extensions.Logging;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace FoundationaLLM.Vectorization.Services.RequestSources
+{
+    public class RequestSourcesBuilder
```
Comment (joelhulen) [inline src/dotnet/Vectorization/Services/RequestSources/RequestSourcesBuilder.cs:16]: Missing XML documentation on class and public properties.

File: src/dotnet/Vectorization/Services/VectorizationService.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+            ValidateRequest(vectorizationRequest);
+
+            var firstRequestSource = _requestSources[vectorizationRequest.Steps.First().Id];
+            await firstRequestSource.SubmitRequest(vectorizationRequest);
+        }
+
+        private static void ValidateRequest(VectorizationRequest vectorizationRequest)
+        {
+            if (vectorizationRequest == null)
+                throw new VectorizationException("The vectorization request should not be null.");
```
Comment (joelhulen) [inline src/dotnet/Vectorization/Services/VectorizationService.cs:31]: Should we log these vectorization exceptions to the logger, or are we expecting the caller to do this?

File: src/dotnet/VectorizationAPI/Program.cs [hunk 6]
```diff
@@ -15,10 +23,17 @@
...
     });
     options.Select("FoundationaLLM:Vectorization:*");
+    options.Select("FoundationaLLM:APIs:VectorizationAPI:*");
 });
 if (builder.Environment.IsDevelopment())
     builder.Configuration.AddJsonFile("appsettings.development.json", true, true);
 
+builder.Services.AddApplicationInsightsTelemetry(new ApplicationInsightsServiceOptions
+{
+    ConnectionString = builder.Configuration["FoundationaLLM:APIs:VectorizationAPI:AppInsightsConnectionString"],
```
Comment (joelhulen) [inline src/dotnet/VectorizationAPI/Program.cs:33]: Consider using the string constants for the App Config keys in `FoundationaLLM.Common.Constants.AppConfigurationKeys`. There are filters and sections within the `AppConfigurationKeyFilters` and `AppConfigurationKeySections` as well. Trying to move away from magic strings where possible and also improve discoverability and documentation through these constants.

File: src/dotnet/Vectorization/Models/VectorizationState.cs (hunk 1 above)
Comment (ciprianjichici) [reply src/dotnet/Vectorization/Models/VectorizationState.cs:79]: Added the XML documentation.

File: src/dotnet/Vectorization/Services/RequestSources/MemoryRequestSourceService.cs (hunk 3 above)
Comment (ciprianjichici) [reply src/dotnet/Vectorization/Services/RequestSources/MemoryRequestSourceService.cs:12]: Added XML documentation.

File: src/dotnet/Vectorization/Services/RequestSources/RequestSourcesBuilder.cs (hunk 4 above)
Comment (ciprianjichici) [reply src/dotnet/Vectorization/Services/RequestSources/RequestSourcesBuilder.cs:16]: Added the XML documentation. While at it, noticed that `VectorizationWorkerBuilder` was also missing the documentation, so I added that as well.

File: src/dotnet/Vectorization/Services/VectorizationService.cs (hunk 5 above)
Comment (ciprianjichici) [reply src/dotnet/Vectorization/Services/VectorizationService.cs:31]: Added local logging for the validation errors. Still throwing though, as it is the responsibility of the caller to provide a valid vectorization request.

File: src/dotnet/VectorizationAPI/Program.cs (hunk 6 above)
Comment (ciprianjichici) [reply src/dotnet/VectorizationAPI/Program.cs:33]: Removed all strings with constant definitions.

**Review by joelhulen — APPROVED** [review]

**Pull request merged**